File: clib/bench/bench.c
#include "../include/sphur.h"
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define NUM_BINS 256
#define NUM_RANDOM 10000

// High-resolution timer
static double get_time_us() {
  struct timespec ts;
  timespec_get(&ts, TIME_UTC);

  return ts.tv_sec * 1e6 + ts.tv_nsec / 1e3;
}

// Benchmark throughput & store numbers in buffer
void benchmark_throughput(sphur_t *state, uint64_t *numbers, size_t n) {
  double start = get_time_us();

  for (size_t i = 0; i < n; i++) {
    numbers[i] = sphur_gen_rand(state); // generate 1 random number
  }

  double end = get_time_us();
  double elapsed_us = end - start;
  double nums_per_us = (double)n / elapsed_us;

  printf("Throughput: %.2f random numbers per microsecond\n", nums_per_us);
}

// Benchmark randomness: histogram & autocorrelation
void benchmark_randomness(uint64_t *numbers, size_t n) {
  // ▶ Histogram / uniformity
  int bins[NUM_BINS] = {0};

  for (size_t i = 0; i < n; i++)
    bins[numbers[i] % NUM_BINS]++;

  double expected = (double)n / NUM_BINS;
  double chi2 = 0.0;

  for (int i = 0; i < NUM_BINS; i++) {
    double diff = bins[i] - expected;
    chi2 += diff * diff / expected;
  }

  printf("Chi-squared (uniformity): %.2f\n", chi2);

  // ▶ Autocorrelation (lag 1)
  double mean = 0.0;

  for (size_t i = 0; i < n; i++)
    mean += numbers[i];

  mean /= n;

  double num = 0.0, den = 0.0;

  for (size_t i = 0; i < n - 1; i++) {
    num += (numbers[i] - mean) * (numbers[i + 1] - mean);
    den += (numbers[i] - mean) * (numbers[i] - mean);
  }

  double autocorr = num / den;
  printf("Autocorrelation (lag 1): %.5f\n", autocorr);
}

// Save numbers to file for Python visualization
void save_numbers_to_file(uint64_t *numbers, size_t n, const char *filepath) {
  FILE *f = fopen(filepath, "w");

  if (!f) {
    perror("Failed to open file");
    exit(1);
  }

  for (size_t i = 0; i < n; i++)
    fprintf(f, "%llu\n", (unsigned long long)numbers[i]);

  fclose(f);
  printf("Random numbers written to: %s\n", filepath);
}

int main() {
  printf("=== Sphur PRNG Benchmark ===\n");

  sphur_t sphur;
  uint64_t seed = 0x9e3779b97f4a7c15ULL;

  if (sphur_init_seeded(&sphur, seed) != 0) {
    printf("Failed to init sphur with seed\n");
    return 1;
  }

  uint64_t *numbers = malloc(NUM_RANDOM * sizeof(uint64_t));

  if (!numbers) {
    printf("Failed to allocate memory\n");
    return 1;
  }

  // Generate numbers & benchmark throughput
  benchmark_throughput(&sphur, numbers, NUM_RANDOM);

  // Benchmark randomness using the same numbers
  benchmark_randomness(numbers, NUM_RANDOM);

  // Save numbers for Python plotting
  save_numbers_to_file(numbers, NUM_RANDOM, "./build/prngs.txt");

  free(numbers);
  printf("Done ✅\n");

  return 0;
}

File: rlib/benchmarks/bench.rs
use sphur::Sphur;
use std::fs::File;
use std::io::Write;
use std::time::Instant;

const NUM_BINS: usize = 256;
const NUM_RANDOM: usize = 10_000;

// Benchmark throughput & store numbers in buffer
fn benchmark_throughput(rng: &mut Sphur, numbers: &mut [u64]) {
    let start = Instant::now();

    for val in numbers.iter_mut() {
        *val = rng.rand().expect("failed to generate rand");
    }

    let elapsed = start.elapsed();
    let elapsed_us = elapsed.as_secs_f64() * 1e6;
    let nums_per_us = numbers.len() as f64 / elapsed_us;

    println!(
        "Throughput: {:.2} random numbers per microsecond",
        nums_per_us
    );
}

// Benchmark randomness: histogram & autocorrelation
fn benchmark_randomness(numbers: &[u64]) {
    // ▶ Histogram / uniformity
    let mut bins = vec![0usize; NUM_BINS];

    for &num in numbers {
        bins[(num % NUM_BINS as u64) as usize] += 1;
    }

    let expected = numbers.len() as f64 / NUM_BINS as f64;
    let chi2: f64 = bins
        .iter()
        .map(|&count| {
            let diff = count as f64 - expected;
            diff * diff / expected
        })
        .sum();

    println!("Chi-squared (uniformity): {:.2}", chi2);

    // ▶ Autocorrelation (lag 1)
    let mean: f64 = numbers.iter().copied().map(|x| x as f64).sum::<f64>() / numbers.len() as f64;

    let mut num = 0.0;
    let mut den = 0.0;

    for i in 0..numbers.len() - 1 {
        let x = numbers[i] as f64 - mean;
        let y = numbers[i + 1] as f64 - mean;
        num += x * y;
        den += x * x;
    }

    let autocorr = num / den;
    println!("Autocorrelation (lag 1): {:.5}", autocorr);
}

// Save numbers to file for Python visualization
fn save_numbers_to_file(numbers: &[u64], filepath: &str) {
    let mut file = File::create(filepath).expect("failed to create output file");

    for &num in numbers {
        writeln!(file, "{}", num).expect("failed to write number");
    }

    println!("Random numbers written to: {}", filepath);
}

fn main() {
    println!("=== Sphur PRNG Benchmark ===");

    let mut rng = Sphur::new_seeded(0x9e3779b97f4a7c15).expect("failed to init sphur with seed");

    let mut numbers = vec![0u64; NUM_RANDOM];

    // Generate numbers & benchmark throughput
    benchmark_throughput(&mut rng, &mut numbers);

    // Benchmark randomness using the same numbers
    benchmark_randomness(&numbers);

    // Save numbers for Python plotting
    save_numbers_to_file(&numbers, "./target/prngs.txt");

    println!("Done ✅");
}
